airflow/providers/amazon/aws/hooks/base_aws.py

```diff
@@ -42,6 +42,7 @@
 import botocore.session
 import requests
 import tenacity
+from aiobotocore.session import AioSession, get_session as async_get_session
```

Comment:
My main concerns do not changed:
- `aiobotocore` pin `botocore` and `boto3` for specific version, last version update takes 7 month, if compare with regular boto stuff it have 3-5 updates per week: fixes, new AWS services and etc. So my opinion that this should not be a part of core dependencies, only if we want to make one or two steps back.
- There was plan to add more consistency (not early than 30 Apr) in hooks and add type hinting by separate existed hooks by https://github.com/apache/airflow/discussions/28560 by separate base hooks. And `aiobotocore.session.AioSession` is not a replacement of `boto3.session.Session`.
- `aiobotocore` can't work with `moto` and some critical stuff tested by `moto`, e.g. auth and AssumeRole
-  My opinion still the same, Airflow core not ready for async stuff `¯\_(ツ)_/¯`, all DB queries are use io blocking implementation (the regular one), so obtain credentials from config still required use some hacks, like `asgiref.sync.sync_to_async` which is transform triggerer service to [some kind of sequential executor](https://github.com/apache/airflow/pull/29038#discussion_r1100693885) 
- Current PR implementation use only blocking io features, can't find any `asyncio` usage.

Personally I would not use veto for this PR, as well as other async/deferrable related stuff. I start to believe that more we add stuff which impact performance degradation more faster we add native asyncio support in core.

Reply:
Having separated base hook for async will not solve the first point you mentioned as even if we have 2 hooks eventually they must agree on boto3 version.

This is one of the cases where the shared governance model of providers is helpful. Lets hear AWS folks perspective and if they considered the issues you raised.
@Taragolis makes good point about the release cycle of aiobotocore and the effect it will have on limiting boto3 version for long period of time

cc @syedahsn @vincbeck @ferruzzi @o-nikolas  @shubham22

Reply:
>    `aiobotocore` pin `botocore` and `boto3` for specific version, last version update takes 7 month, if compare with regular boto stuff it have 3-5 updates per week: fixes, new AWS services and etc. So my opinion that this should not be a part of core dependencies, only if we want to make one or two steps back.
> 
That 7 month window is kind of the worst case. I'm seeing aiobotocore releases which bump botocore as short as one month (and everything in between). 
>There was plan to add more consistency (not early than 30 Apr) in hooks and add type hinting by separate existed hooks by [Amazon Provider: Consistency in boto3-based Hooks #28560](https://github.com/apache/airflow/discussions/28560) by separate base hooks. And `aiobotocore.session.AioSession` is not a replacement of `boto3.session.Session`.

Agree with @eladkal on this one.
>`aiobotocore` can't work with `moto` and some critical stuff tested by `moto`, e.g. auth and AssumeRole

I personally haven't looked into this. But, sure, we may have to do more manual mocking and testing for the aio usecase.
 
>My opinion still the same, Airflow core not ready for async stuff `¯\_(ツ)_/¯`, all DB queries are use io blocking implementation (the regular one), so obtain credentials from config still required use some hacks, like `asgiref.sync.sync_to_async` which is transform triggerer service to [some kind of sequential executor](https://github.com/apache/airflow/pull/29038#discussion_r1100693885)

Honestly, I think this ship has long since sailed. I agree that the architecture of deferrable operators/trigger is a bit clunky, leads to a lot of code duplication, and has this duality of sync vs async hooks. But the fact of the matter is Deferrable Operator support is public and we hear from users and customers that they want to use it. So I think an incremental/agile approach is often the best way to approach it. If we wait until there is perfection, we'd never ship anything.

Maybe a helpful exercise would be to list some concrete examples of how this PR would degrade an Airflow user's experience more than it would improve it (other than the botocore version, that one is quite clear).

Reply:
First of all I think we need to be pragmatic and find some reasonable tacktical solutions, even if they are not perfect. And possibly work on strategic solutions to solve the problems in a long term.

TL;DR; We seem to have to make some compromises for now. I think the best compromise it so allow async operators to be merged, while limiting to what we can release for AWS provider (also async) limited what pinned botocore version allows us to do.

I understand @Taragolis concerns and they are very sound. The problem is that we have no easy way out of some  of those problems  in a short term, so the question is what we can do now and whether it will block us to fix it in the long term (if possible).

> aiobotocore pin botocore and boto3 for specific version

This is unfortunate but until AWS will decide to build async support into botocore, I think we will  have to rely on aiobotocore playing the catch-up game. The thing with botocore is that it is pushed frequently, but in fact our users are not updating that frequently, so as long as what we are ok with limiting ourselves in what we publish in AWS operators to whatever version of botocore is pinned, this should be acceptable. This means that new features will have to wait for upgrades of aiobotocore (for example latest release of aiobotocore from a week ago enabled https://github.com/apache/airflow/pull/28850#issuecomment-1467791465 to be merged).  If the AWS team is fine for it, this is fine for me. We often have delays in updating some dependencies and this is quite OK. 
 
Users could use PythonVirtual or External operator to use botocore's new features - more hassle, but I guess this is pretty acceptable solution for anything that is not available in the "regular" operators - and we should likely describe it somewhere. 

Also pinning the botocore in this case in our CI will be beneficial for that approach - because we will have only the "compatible" version of botocore installed. 
 
Question to @o-nikolas and the taem - maybe it is already planned or maybe you could raise it and push internally to add async support directly in boto library? That sounds like something that Amazon will be presurred to do anyway I guess? That would be a long-term strategic solution if we know at least that it is coming at some point in time.

> There was plan to add more consistency

I have not looked into many details of the AWS hook split - but, it looks good, and I think it can be taken and treated separately (it does not solve indeed botocore/aiobotocore pinning). The question is - is there any problem with joing the two approaches @Taragolis ? Can we continue the split while using either of the sessions? AioSession derives from  botocore.Session, so while (minus pinning problems), it seems doable to define our own session essentially (guarded by TYPE_CHECKING dynamically) which would be boto.session (if no aiocore installed) or `AioSession | Session` (if installed) and that would actually help with developing providers that will be compatible with the pinned botocore (through our CI).

> aiobotocore can't work with moto

Yeah. more mocking in-house should be needed. Agree with @o-nikolas . Price to pay for async.

> Airflow core not ready for async stuff

I think this is a valid concern and not everything is perfect. And yes, it is a time-bomb potentially. I would like however to have @andrewgodwin comment on that - but I believe we use `sync_to_async` where necessary (same as in django) for places that are not ready and while it does introduce some serialization it is not catastrophic and is a good way to add **mostly** async support. I think Andrew implemented (or at least participated in it) in django for similar cases and maybe we can hear from him a comment about the currnet sync_to_async use. Maybe we could also (based on the comments) start figuring out a long term strategy on how to deal with it.

I think if we agree on this direction, it will unblock tactical changes to add more deferrable operators for AWS, while we could also work on improving the infrastructure and (hopefully) get to the point where async support will be provided by botocore maybe? 

I don't see any other constructive way to approach it, to be honest.

Reply:
Thanks for looking at the PR! You raised some good points:

The versioning of aiobotocore is a valid issue. However, as @o-nikolas pointed out, the 7 month gap seems to be an outlier. The aiobotocore library has a fairly active active community that regularly releases newer versions. Also, while botocore gets updated regularly, generally, there aren't major changes between each update, so while it seems like aiobotocore may pin the botocore version, it shouldn't have major impact on user functionality. 
>There was plan to add more consistency (not early than 30 Apr) in hooks and add type hinting by separate existed hooks by https://github.com/apache/airflow/discussions/28560 by separate base hooks. And aiobotocore.session.AioSession is not a replacement of boto3.session.Session.

This is a really cool idea, and I'm not certain where aiobotocore would fit in with this paradigm. But while you are correct that `aiobotocore.session.AioSession` is not a replacement to a `boto3.session.Session`, I think that because they both share the botocore library, there are enough similarities between the 2 libraries that the `AioSession` can fit into that model.

>all DB queries are use io blocking implementation (the regular one)

The purpose of this PR was to allow workers to be freed up when involved with time consuming network tasks such as polling services for a certain state etc. DB queries are io blocking operations, and we can apply those optimizations in Amazon Provider once we have a precedence for applying it across Airflow. I believe Network operation optimizations are themselves a big step forward.

Reply:
> Question to @o-nikolas and the taem - maybe it is already planned or maybe you could raise it and push internally to add async support directly in boto library? That sounds like something that Amazon will be presurred to do anyway I guess? That would be a long-term strategic solution if we know at least that it is coming at some point in time.

We discussed this with the Boto team last October when we added deferrable operators to our roadmap. Yesterday, we revisited the topic with them. The Boto team is already planning to include async functionality in a major release, which will eliminate the need for the aiobotocore library entirely. However, this release won't be available before 2024 as it will be a major change. This is why we decided to use aiobotocore in the short-term as it is the best way forward, even though it's not officially endorsed by the AWS Boto team.

Reply:
>so while it seems like aiobotocore may pin the botocore version, it shouldn't have major impact on user functionality.

But what about boto3 version pin?

Reply:
> But what about boto3 version pin?

My view: Assuming that aiobotocore will be released semi-frequently and catch-up with few months old version of boto3/botocore the worst that can happen is some new features/fixes will only be available with few months delay (when new version of aiobotocore is released and we release next version of the provider). Which is not that far from expectations of the users IMHO. Also if - for example - AWS team will see the need of acceleration of an upgrade, there is always possibility of not only opening an issue to aiobotocore team, but also making PR to update to the latest API/features (though it would likely require more involvement and learning).

The users will always have an option to install newer boto3 version in Virtualenv/External or use DockerOperator - following https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#handling-conflicting-complex-python-dependencies (and we could add docs explaining this). 

Also airflow itself and even provider will not limit the boto3. The constraints are "golden" set of dependencies and if someone would like to use newer boto3 and not use aiobotocore extra - it will be possible for them to upgrade  boto (and we can also document). There will be no "check" that the newer version will work - as we will not run tests for it - but the user will be able to do their own tests and use newer boto. We will not prevent that. 

I don't think we can solve all issues here - it's impossible. We have to sacrifice something. This approach is not perfect, but at least we might have steady progress with new deferrable operators for AWS being added.

Reply:
> if someone would like to use newer boto3 and not use aiobotocore extra - it will be possible for them to upgrade boto (and we can also document). There will be no "check" that the newer version will work - as we will not run tests for it - but the user will be able to do their own tests and use newer boto. We will not prevent that.

I'm cool with this. I'm just saying that there is also the option of the other way around.
Till native and conflictless approach introduced, maybe async operators should be external in their own dedicated package and to be installed by users who seek to use triggers? That way current behavior and expectations are not sacrificed and who ever wants the new functionality can install it like any 3rd party provider.

I'm worried that we may underestimate the impact of limiting boto3.

> We have to sacrifice something

I'm OK with any decision on this. I just want to make sure that all aspects were taken into account.

Reply:
> Till native and conflictless approach introduced, maybe async operators should be external in their own dedicated package and to be installed by users who seek to use triggers? 

That is also a possibility of course. The end result is though, that if in the "sync" package you would like to use some new feature you technically should use ">=version_when_it_was_added".

Byt you really can't add such limit - If you add such limit in the "sync" package, then you won't be able to install those two packages together (which I guess is out-of-the-question). Effectively this means that we will cause breaking some "sync" package features by installing the "async" package. Not even mentioning installing "older" version of the "async" package with newer version of "sync" one - this will break even more things and quite unpredictably. I already imagine the issues raised by the users :).

It will also cause quite a lot of duplication of the code which necessarily will have to be in both packages. I think sharing a code between two packages like that will be quite complex. IMHO it's far more complex to maintain, fix problems, make sure that there is some level of compatibility and handling errors will be a nightmare to maintain. 

I on the other hand, I am worried we may underestimate complexities of maintaining such split :) . Effectively we take the burden on the maintainers/ aws team to handle all the complexity involved rather than leave it up to the users to slightly complicate their workflows if they want to be on the "bleeding edge". 

But if AWS team is ok to maintain such a split -  I am ok with this scenario as well - it is certainly doable. Adding new `asyncaws` provider could be done. Maybe even extracting `common.aws` if we would like to share the code and those who do it will not be afraid of handling cases like we had with common.sql - I just believe it will be super-complex to maintain. You've been warned :).

Reply:
Sorry for late response, do not have time for Airflow project now, and even do not use it in production right now. So better say I'm not a user of Airflow right now.

I would try to to answer for some messages what I miss.

Note: Everything just my opinion, and I wouldn't try to change anyone opinion mind, because I found on real experience that this not work and all changes request could be removed   

> That 7 month window is kind of the worst case. I'm seeing aiobotocore releases which bump botocore as short as one month (and everything in between).

I wouldn't so optimistic about `aiobotocore` releases, all `botocore` bumps since late 2017.

| Bump Date    | Version  | Days Since LU | Months Since LU |
|--------------|----------|--------------:|----------------:|
| Mar 7, 2023  | 1.29.76  |           225 |             7,4 |
| Aug 25, 2022 | 1.27.59  |           130 |             4,3 |
| Mar 17, 2022 | 1.24.21  |            93 |             3,1 |
| Dec 14, 2021 | 1.23.24  |            42 |             1,4 |
| Nov 2, 2021  | 1.22.8   |           294 |             9,7 |
| Jan 12, 2021 | 1.19.52  |           178 |             5,9 |
| Aug 18, 2020 | 1.17.44  |           148 |             4,9 |
| Feb 21, 2020 | 1.15.3   |           101 |             3,3 |
| Nov 12, 2019 | 1.13.14  |            18 |             0,6 |
| Oct 25, 2019 | 1.12.252 |            99 |             3,3 |
| Jul 18, 2019 | 1.12.189 |           159 |             5,2 |
| Feb 9, 2019  | 1.12.91  |            62 |               2 |
| Dec 9, 2018  | 1.12.49  |           143 |             4,7 |
| Jul 19, 2018 | 1.10.58  |            75 |             2,5 |
| May 5, 2018  | 1.10.12  |           124 |             4,1 |
| Jan 1, 2018  | 1.8.21   |             0 |               0 |
| Jan 1, 2018  | 1.8.20   |             0 |               0 |
| Jan 1, 2018  | 1.8.12   |             0 |               0 |
| Jan 1, 2018  | 1.7.48   |            47 |             1,5 |
| Oct 15, 2017 | 1.7.27   |           n/a |             n/a |

So if exclude multiple bumps in 2018-01-01 than optimistic scenario 4 bumps per year, pessimistic 2 bumps per year and 3.7 median delay per update.

> I personally haven't looked into this. But, sure, we may have to do more manual mocking and testing for the aio usecase.

In past we have fully mocked version of auth as result: https://github.com/apache/airflow/pull/27198 which we found only after 2 months, and affected Amazon Provider 5.1.0 and 6.0.0 (bundled in AWS MWAA 2.4.3)

>  Honestly, I think this ship has long since sailed. I agree that the architecture of deferrable operators/trigger is a bit clunky, leads to a lot of code duplication, and has this duality of sync vs async hooks. But the fact of the matter is Deferrable Operator support is public and we hear from users and customers that they want to use it. So I think an incremental/agile approach is often the best way to approach it. If we wait until there is perfection, we'd never ship anything.

Potentially users doesn't know about all side effects of Deferrable Operators, because it hided from them and presented as better that sync implementation.
To be honest we do not have any comparison of Sync vs Deferrable for consumption resources, latency, and average time to execute.

> Maybe a helpful exercise would be to list some concrete examples of how this PR would degrade an Airflow user's experience more than it would improve it (other than the botocore version, that one is quite clear).

This PR only uses blocking io implementation, that mean event loop of triggerer would be blocked.

> I have not looked into many details of the AWS hook split - but, it looks good, and I think it can be taken and treated separately (it does not solve indeed botocore/aiobotocore pinning). The question is - is there any problem with joing the two approaches @Taragolis ?

My opinion is that async for Amazon Provider should be Optional dependency which better live separate of Sync implementation:
- Async will not supported all parameters of AWS Connection
- `boto3.client` in most cases is basically `botocore.{SomeSpecificClient}`, except `AWS S3`
- Dependency hell if user want to upgrade `boto3`, that back to the past when `boto3` and `botocore` have weird dependency policy and make `pip` resolve crazy

> I think this is a valid concern and not everything is perfect. And yes, it is a time-bomb potentially. I would like however to have @andrewgodwin comment on that - but I believe we use sync_to_async where necessary (same as in django) for places that are not ready and while it does introduce some serialization it is not catastrophic and is a good way to add mostly async support.

![image](https://user-images.githubusercontent.com/3998685/225137845-0063d24a-1868-49da-8030-123c9275fd0e.png)

There were PR's which initially tried almost everything wrap to `sync_to_async`. I do not check all of them maybe some of them merged into some provider or core, I've only known about this ones:
- https://github.com/apache/airflow/pull/29260
- https://github.com/apache/airflow/pull/29313

I'm more unhappy that we do not disclose information to users about all side effect of `sync_to_async` deferrable operators implementation, and this affect all of them except `TimeSensorAsync` and `TimeDeltaSensorAsync`

Reply:
I agree. I don't have a good answer here. This is a dilemma.
any course of action taken will have pros and cons.

I'm fine with either way we go. We raised the questions I think now we should leave it to AWS folks to debate and let us know their chosen path (in the spirit of shared governance model)

Reply:
The aiobotocore dependency was merged in a separate PR and combined with Jarek's new testing we should be in a pretty good spot as far as compromise on this front.
I'm resolving this discussion, please re-open if there is anything else to discuss on the topic!